File: backend/models/farm.go
package models
import (
	"time"
	"github.com/google/uuid"
	"gorm.io/gorm"
)
type Farm struct {
	ID          uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	UserID      uuid.UUID      `gorm:"type:uuid;not null;index" json:"user_id"`
	Name        string         `gorm:"type:varchar(255);not null" json:"name" validate:"required,min=2,max=255"`
	Location    string         `gorm:"type:text" json:"location"`
	Area        float64        `gorm:"type:decimal(10,2)" json:"area"`
	Description string         `gorm:"type:text" json:"description"`
	Image       string         `gorm:"type:varchar(500)" json:"image"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	User  User   `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Ponds []Pond `gorm:"foreignKey:FarmID" json:"ponds,omitempty"`
}
func (f *Farm) BeforeCreate(tx *gorm.DB) error {
	if f.ID == uuid.Nil {
		f.ID = uuid.New()
	}
	return nil
}
type Pond struct {
	ID        uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	FarmID    uuid.UUID      `gorm:"type:uuid;not null;index" json:"farm_id"`
	Name      string         `gorm:"type:varchar(255);not null" json:"name" validate:"required,min=2,max=255"`
	Area      float64        `gorm:"type:decimal(10,2)" json:"area"`
	Depth     float64        `gorm:"type:decimal(5,2)" json:"depth"`
	Status    string         `gorm:"type:varchar(50);default:'active'" json:"status"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
	Farm              Farm               `gorm:"foreignKey:FarmID" json:"farm,omitempty"`
	CultivationCycles []CultivationCycle `gorm:"foreignKey:PondID" json:"cultivation_cycles,omitempty"`
}
func (p *Pond) BeforeCreate(tx *gorm.DB) error {
	if p.ID == uuid.Nil {
		p.ID = uuid.New()
	}
	return nil
}

File: fluttermultiplatform/lib/models/farm.dart
import 'pond.dart';
import 'user.dart';

class Farm {
  final String id;
  final String userId;
  final String name;
  final String location;
  final double area;
  final String description;
  final String image;
  final List<Pond>? ponds;
  final User? user;
  final String createdAt;

  Farm({required this.id, this.userId = '', required this.name, this.location = '', this.area = 0, this.description = '', this.image = '', this.ponds, this.user, this.createdAt = ''});

  factory Farm.fromJson(Map<String, dynamic> json) => Farm(
        id: json['id']?.toString() ?? '',
        userId: json['user_id']?.toString() ?? '',
        name: json['name'] ?? '',
        location: json['location'] ?? '',
        area: (json['area'] ?? 0).toDouble(),
        description: json['description'] ?? '',
        image: json['image'] ?? '',
        ponds: json['ponds'] != null ? (json['ponds'] as List).map((e) => Pond.fromJson(e)).toList() : null,
        user: json['user'] != null ? User.fromJson(json['user']) : null,
        createdAt: json['created_at'] ?? '',
      );
}
